client/src/details/SFSUrlBuilder.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "SFSUrlBuilder.h"
+
+using namespace SFS::details;
+
+constexpr const char* c_apiVersion = "v2";
+constexpr const char* c_apiDomain = "api.cdp.microsoft.com";
+
+std::unique_ptr<SFSUrlBuilder> SFSUrlBuilder::CreateFromAccountId(const std::string& accountId,
+                                                                  std::string instanceId,
+                                                                  std::string nameSpace,
+                                                                  const ReportingHandler& handler)
+{
+    auto builder =
+        std::unique_ptr<SFSUrlBuilder>(new SFSUrlBuilder(std::move(instanceId), std::move(nameSpace), handler));
+    builder->SetScheme(Scheme::Https);
+    builder->SetHost(accountId + "." + std::string(c_apiDomain));
+    return builder;
+}
+
+std::unique_ptr<SFSUrlBuilder> SFSUrlBuilder::CreateFromCustomUrl(const std::string& customUrl,
+                                                                  std::string instanceId,
+                                                                  std::string nameSpace,
+                                                                  const ReportingHandler& handler)
+{
+    auto builder =
+        std::unique_ptr<SFSUrlBuilder>(new SFSUrlBuilder(std::move(instanceId), std::move(nameSpace), handler));
+    builder->SetUrl(customUrl);
+    return builder;
+}
+
+SFSUrlBuilder::SFSUrlBuilder(std::string&& instanceId, std::string&& nameSpace, const ReportingHandler& handler)
+    : UrlBuilder(handler)
+    , m_instanceId(std::move(instanceId))
+    , m_nameSpace(std::move(nameSpace))
+{
+}
+
+std::string SFSUrlBuilder::GetLatestVersionUrl(const std::string& product)
+{
+    SetPath(GetVersionsUrlPath(product) + "latest");
+    SetQuery("action=select");
+    return GetUrl();
+}
+
+std::string SFSUrlBuilder::GetLatestVersionBatchUrl()
+{
+    SetPath(GetNamesUrlPath());
+    SetQuery("action=BatchUpdates");
+    return GetUrl();
+}
+
+std::string SFSUrlBuilder::GetSpecificVersionUrl(const std::string& product, const std::string& version)
+{
+    SetPath(GetVersionsUrlPath(product) + EscapeString(version));

```

Comment:
Why not add an AppendPath method to UrlBuilder and use that to build the path? Will that also allow you to let curl do the url escaping?

This is more readable IMO, especially for more complex ones like in `GetNamesUrlPath`:
`SetPath(GetVersionsUrlPath(product), true /*encode*/);`
`AppendPath(version, true /*encode*/);`
`AppendPath(files, true /*encode*/);`

Reply:
Curl doesn't have a facility for this, but we can implement a custom one